pydantic/schema.py

```diff
@@ -723,42 +748,79 @@ def encode_default(dft: Any) -> Any:
 
     :param annotation: an annotation from a field specification, as ``str``, ``ConstrainedStr``
     :param field_info: an instance of FieldInfo, possibly with declarations for validations and JSON Schema
     :return: the same ``annotation`` if unmodified or a new annotation with validation in place
     """
-    if isinstance(annotation, type):
-        attrs: Optional[Tuple[str, ...]] = None
-        constraint_func: Optional[Callable[..., type]] = None
-        if issubclass(annotation, str) and not issubclass(annotation, (EmailStr, AnyUrl, ConstrainedStr)):
-            attrs = ('max_length', 'min_length', 'regex')
-            constraint_func = constr
-        elif lenient_issubclass(annotation, numeric_types) and not issubclass(
-            annotation, (ConstrainedInt, ConstrainedFloat, ConstrainedDecimal, ConstrainedList, bool)
-        ):
-            # Is numeric type
-            attrs = ('gt', 'lt', 'ge', 'le', 'multiple_of')
-            numeric_type = next(t for t in numeric_types if issubclass(annotation, t))  # pragma: no branch
-            constraint_func = _map_types_constraint[numeric_type]
-        elif issubclass(annotation, ConstrainedList):
-            attrs = ('min_items', 'max_items')
-            constraint_func = conlist
-        if attrs:
-            kwargs = {
-                attr_name: attr
-                for attr_name, attr in ((attr_name, getattr(field_info, attr_name)) for attr_name in attrs)
-                if attr is not None
-            }
-            if kwargs:
-                constraint_func = cast(Callable[..., type], constraint_func)
-                return constraint_func(**kwargs)
+    item_type = get_item_type(annotation)
+    if not isinstance(item_type, type):
+        return annotation
+
+    attrs: Tuple[str, ...] = ()
+    constraint_func: Optional[Callable[..., type]] = None
+    if lenient_issubclass(annotation, ConstrainedList):
+        attrs = ('min_items', 'max_items')
+        constraint_func = conlist
+    elif issubclass(item_type, str) and not issubclass(item_type, (EmailStr, AnyUrl, ConstrainedStr)):
+        attrs = ('max_length', 'min_length', 'regex')
+        constraint_func = constr
+    elif issubclass(item_type, bytes):
+        attrs = ('max_length', 'min_length', 'regex')
+        constraint_func = conbytes
+    elif issubclass(item_type, numeric_types) and not issubclass(
+        item_type, (ConstrainedInt, ConstrainedFloat, ConstrainedDecimal, ConstrainedList, bool)
+    ):
+        # Is numeric type
+        attrs = ('gt', 'lt', 'ge', 'le', 'multiple_of')
+        numeric_type = next(t for t in numeric_types if issubclass(item_type, t))  # pragma: no branch
+        constraint_func = _map_types_constraint[numeric_type]
+
+    unused_constraints = [
+        f for f in validation_attribute_to_schema_keyword if f not in attrs and getattr(field_info, f) is not None
+    ]
+    if unused_constraints:
```

Comment:
@dmontagu @tiangolo I need your feedback/advice on this.

I think we should tell the user if they add constraints but they're not applied, however are there situations where a custom could set a constraint which would show in schema, but not be applied here? I couldn't find any in tests, but that doesn't mean there aren't any.

If so we might need a config argument or extra argument to `Field()` to skip this check.

Reply:
I can't think of any, but it feels a little strong to me to raise an error rather than a warning or having a flag to allow it or similar.

That said, specifying `schema_extra` would work as a way to add these to the schema without validation if you really wanted to right?

If that's right, I think this is fine, but maybe it would be worth calling this out in the error message?

-------

More generally, I think it would be nice if it were easier to implement the `constraint_func` for the class and have pydantic pick it up, so you could create custom classes with custom validation as a function of the specified schema. But that is probably outside the scope of what should be handled now.

Reply:
> More generally, I think it would be nice if it were easier to implement the constraint_func for the class and have pydantic pick it up, so you could create custom classes with custom validation as a function of the specified schema. But that is probably outside the scope of what should be handled now.

Surely that should be done via type annotations?

> If that's right, I think this is fine, but maybe it would be worth calling this out in the error message?

Yes agreed.

Reply:
after the changes I've just made, I can't think of a scenario where constraints would be valid in schema, but not enforced.

To add an example to the docs and an explanation of the work around we need at least one example where this error would be incorrectly raised :-)

Reply:
I don't have strong opinions about this.

I see how it could feel cumbersome if it suddenly raises errors for models that didn't before. 

But at the same time, I can imagine someone expecting it to show validation errors for some data with constraints in `Field` that are not actually applied. And I can see how they could forget to check the warnings on the logs...

---

Something I can think of to mitigate these use cases is, in the docs, add an alternative with pure `Field` for those cases, I suggested it in the PR review.

Another idea that comes to mind is to show a warning first as @dmontagu suggests, and make it a deprecation warning. And in some next version, throw an error. Just an idea, not that I think that's necessary, both approaches seem fine to me.